trackpy/linking.py

```diff
@@ -744,139 +768,231 @@ def link_iter(levels, search_range, memory=0,
+                              self.neighbor_strategy)
 
-        # Sort out what can go to what.
-        assign_candidates(cur_level, prev_hash, search_range,
-                          neighbor_strategy)
+            # sort the candidate lists by distance
+            for p in cur_set:
+                p.back_cands.sort(key=lambda x: x[1])
+            for p in prev_set:
+                p.forward_cands.sort(key=lambda x: x[1])
 
-        # sort the candidate lists by distance
-        for p in cur_set:
-            p.back_cands.sort(key=lambda x: x[1])
-        for p in prev_set:
-            p.forward_cands.sort(key=lambda x: x[1])
+            # Note that this modifies cur_set, prev_set, but that's OK.
+            spl, dpl = self.assign_links(cur_set, prev_set, self.search_range)
 
-        new_mem_set = set()
-        # while there are particles left to link, linkge the repo
-        while len(cur_set) > 0:
-            p = cur_set.pop()
+            new_mem_set = set()
+            for sp, dp in zip(spl, dpl):
+                # Do linking
+                if sp is not None and dp is not None:
+                    sp.track.add_point(dp)
+                    _maybe_remove(self.mem_set, sp)
+                elif sp is None:
+                    # if unclaimed destination particle, a track is born!
+                    self.track_lst.append(self.track_cls(dp))
+                elif dp is None:
+                    # add the unmatched source particles to the new
+                    # memory set
+                    new_mem_set.add(sp)
+
+                # Clean up
+                if dp is not None:
+                    del dp.back_cands
+                if sp is not None:
+                    try:
+                        del sp.forward_cands
+                    except AttributeError:
+                        pass
+
+            # set prev_hash to cur hash
+            prev_hash = cur_hash
+
+            # add in the memory points
+            # store the current level for use in next loop
+            if self.memory > 0:
+                # identify the new memory points
+                new_mem_set -= self.mem_set
+                mem_history.append(new_mem_set)
+                # remove points that are now too old
+                self.mem_set -= mem_history.pop(0)
+                # add the new points
+                self.mem_set |= new_mem_set
+                # add the memory particles to what will be the next source
+                # set
+                tmp_set |= self.mem_set
+                # add memory points to prev_hash (to be used as the next source)
+                for m in self.mem_set:
+                    # add points to the hash
+                    prev_hash.add_point(m)
+                    # re-create the forward_cands list
+                    m.forward_cands = []
+
+            prev_set = tmp_set
+
+            yield cur_level
+
+    def assign_links(self, dest_set, source_set, search_range,
```

Comment:
As this is coming back to me, it needs to mutate the set as that is how memory works.  Should this be a private method ?